### ISSUE TITLE
Minor problem from #252

### ISSUE THREAD

**Issue by Alarson93** [evermeer/EVReflection#260]
Hello again! It would seem I had a minor problem in my proposed fix in #252 

In the `open override func setValue(_ value: Any!, forUndefinedKey key: String)` we have `else if type(of: current) == type(of: value)`.
The `type(of: value)` returns `Swift.Optional<Any>`. If we force value using `type(of: value!)`, we get `RealmSwift.List<Swift.String>` as expected. Forcing isn't dangerous in this case because the function already forces the value parameter.

The function, as a whole, including the force unwrap of optional I mentioned would like so:

```
    open override func setValue(_ value: Any!, forUndefinedKey key: String) {
        if let kvc = self as? EVGenericsKVC {
            kvc.setGenericValue(value as AnyObject!, forUndefinedKey: key)
        } else if let current = super.value(forKey: key) {
            if current is NSNumber, value is NSString {
                //If we do have key and the types are different, then this is called from EVReflection
                let num = NSNumber(value: Int("\(value as! NSString)") ?? 0)
                super.setValue(num, forUndefinedKey: key)
            } else if type(of: current) == type(of: value!)
                || (String(describing: type(of: current)).hasPrefix("List<") && value! is NSArray) {
                super.setValue(value, forUndefinedKey: key)
            } else {
                (self as? EVReflectable)?.addStatusMessage(.InvalidType, message: "Invalid type for the key '\(key)' in the class '\(EVReflection.swiftStringFromClass(self))'")
                evPrint(.IncorrectKey, "WARNING: the type (\(type(of: current)) of the The key '\(key)' in class '\(EVReflection.swiftStringFromClass(self))' does not corresponds to the type (\(type(of: value!)) in the json or Realm database.\n")
            }
        } else if objectSchema.properties.first(where: { $0.name == key }) != nil {
            super.setValue(value, forUndefinedKey: key)
        } else {
            (self as? EVReflectable)?.addStatusMessage(.IncorrectKey, message: "The class '\(EVReflection.swiftStringFromClass(self))' is not key value coding-compliant for the key '\(key)'")
            evPrint(.IncorrectKey, "WARNING: The class '\(EVReflection.swiftStringFromClass(self))' is not key value coding-compliant for the key '\(key)'\n❓ This could be a strange Realm List issue where the key is reported undefined but it's still set.\n")
        }
    }
```

Tested and working.
 

TL;DR - I just need to force unwrap "value" when checking it's type. `type(of: value!)`

**Comment by evermeer** [comment]
Thanks! I have added the change to version 5.4.3
The publish to cocoapods is currently running...

**Comment by wadevanne** [comment]
Hi! I maybe had some problem just like that, and it seems still not working with properties added in an object. 
It crashes at line 24 in RealmObjectExtension.swift
```
} else if let current = super.value(forKey: key) { 
```
line 173 in RLMObjectBase.mm
```
return [super valueForKey:key]; 
```
line 181 in RLMObjectBase.mm
```
return [super valueForUndefinedKey:key];
```
Now, I just override function setValue(_:forUndefinedKey:) without calling its super to avoid this.

It's like we had two properties in OneObject save at realm
```
@objcMembers
open class OneObject: Object {
    public dynamic var string: String?
    public dynamic var address: Address?
}
```
and now the JSON string from server updated, we should also update OneObject like:
```
open class OneObject: Object {
    // public dynamic var string: String? // no more
    public dynamic var address: Address?
    public dynamic var newString: String?
    public dynamic var count: Int = 0
}
```
'newString' and 'count' are the added undefined key
'string' is removed

- Here is the problem:

before updating OneObject and release a new version published at App Store
we execute function ‘realm add', it will crashed at the undefined key no matter which one.

BTW, it works well when the data from server removed properties only.

**Comment by evermeer** [comment]
I will have a look at this.

**Comment by wadevanne** [comment]
Thanks!!! It works well now and shows the great warning tips about undefined keys as it always do.
